When targeting earlier versions of macOS, Xcode will complain about `getentropy`'s availability

In some configurations (such as ours) where `-Werror=unguarded-availability` is set, compiling Cryptography results in the following compile error:
```
build/temp.macosx-10.14-x86_64-3.7/_openssl.c:1819:10: error: 'getentropy' is only available on macOS 10.12 or newer [-Werror,-Wunguarded-availability]
    if (&getentropy != NULL) {
         ^~~~~~~~~~
build/temp.macosx-10.14-x86_64-3.7/_openssl.c:635:16: note: 'getentropy' has been marked as being introduced in macOS 10.12 here, but the deployment target is macOS 10.10.0
    extern int getentropy(void *buffer, size_t size) __attribute((weak_import));
               ^
build/temp.macosx-10.14-x86_64-3.7/_openssl.c:1819:10: note: enclose 'getentropy' in a __builtin_available check to silence this warning
    if (&getentropy != NULL) {
         ^~~~~~~~~~
build/temp.macosx-10.14-x86_64-3.7/_openssl.c:1847:19: error: 'getentropy' is only available on macOS 10.12 or newer [-Werror,-Wunguarded-availability]
            res = getentropy(buffer, (size_t)len);
                  ^~~~~~~~~~
build/temp.macosx-10.14-x86_64-3.7/_openssl.c:635:16: note: 'getentropy' has been marked as being introduced in macOS 10.12 here, but the deployment target is macOS 10.10.0
    extern int getentropy(void *buffer, size_t size) __attribute((weak_import));
               ^
build/temp.macosx-10.14-x86_64-3.7/_openssl.c:1847:19: note: enclose 'getentropy' in a __builtin_available check to silence this warning
            res = getentropy(buffer, (size_t)len);
                  ^~~~~~~~~~
```

Apple recommends the use of `__builtin_available` to runtime-check whether a function is available, so it complains. Cryptography currently uses `weak_import` and checks whether the symbol is `NULL`, which achieves the same result, so this should be silenced.

--- a/src/_cffi_src/openssl/src/osrandom_engine.c
+++ b/src/_cffi_src/openssl/src/osrandom_engine.c
@@ -251,7 +251,13 @@ static int osrandom_init(ENGINE *e) {
 #if !defined(__APPLE__)
     getentropy_works = CRYPTOGRAPHY_OSRANDOM_GETENTROPY_WORKS;
 #else
+
+    // clang will complain that use of `getentropy` is not guarded, but the
+    // symbol is already weakly imported
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability"
     if (&getentropy != NULL) {
+#pragma clang diagnostic pop
         getentropy_works = CRYPTOGRAPHY_OSRANDOM_GETENTROPY_WORKS;
     } else {
         getentropy_works = CRYPTOGRAPHY_OSRANDOM_GETENTROPY_FALLBACK;
@@ -277,7 +283,10 @@ static int osrandom_rand_bytes(unsigned char *buffer, int size) {
         while (size > 0) {
             /* OpenBSD and macOS restrict maximum buffer size to 256. */
             len = size > 256 ? 256 : size;
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunguarded-availability"
             res = getentropy(buffer, (size_t)len);
+#pragma clang diagnostic pop
             if (res < 0) {
                 ERR_Cryptography_OSRandom_error(
                     CRYPTOGRAPHY_OSRANDOM_F_RAND_BYTES,

--- a/src/_cffi_src/openssl/src/osrandom_engine.c
+++ b/src/_cffi_src/openssl/src/osrandom_engine.c
@@ -251,13 +251,7 @@ static int osrandom_init(ENGINE *e) {
 #if !defined(__APPLE__)
     getentropy_works = CRYPTOGRAPHY_OSRANDOM_GETENTROPY_WORKS;
 #else
-
-    // clang will complain that use of `getentropy` is not guarded, but the
-    // symbol is already weakly imported
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wunguarded-availability"
-    if (&getentropy != NULL) {
-#pragma clang diagnostic pop
+    if (__builtin_available(macOS 10.12, *)) {
         getentropy_works = CRYPTOGRAPHY_OSRANDOM_GETENTROPY_WORKS;
     } else {
         getentropy_works = CRYPTOGRAPHY_OSRANDOM_GETENTROPY_FALLBACK;
@@ -283,6 +277,7 @@ static int osrandom_rand_bytes(unsigned char *buffer, int size) {
         while (size > 0) {
             /* OpenBSD and macOS restrict maximum buffer size to 256. */
             len = size > 256 ? 256 : size;
+/* on mac, availability is already checked using `__builtin_available` above */
 #pragma clang diagnostic push
 #pragma clang diagnostic ignored "-Wunguarded-availability"
             res = getentropy(buffer, (size_t)len);

